OsmAnd-java/src/main/java/net/osmand/router/TransportRoutePlanner.java

```diff
@@ -54,40 +58,57 @@ public List<TransportRouteResult> buildRoute(TransportRoutingContext ctx, LatLon
 			r.distFromStart = r.walkDist / ctx.cfg.walkSpeed;
 			queue.add(r);
 		}
-		
-		double finishTime = ctx.cfg.maxRouteTime;
+
+		Map<Long, Double> finishTimeMap = new HashMap<>();
 		ctx.finishTimeSeconds = ctx.cfg.finishTimeSeconds;
 		if (totalDistance > ctx.cfg.maxRouteDistance && ctx.cfg.maxRouteIncreaseSpeed > 0)  {
 			int increaseTime = (int) ((totalDistance - ctx.cfg.maxRouteDistance) 
 					* 3.6 / ctx.cfg.maxRouteIncreaseSpeed);
-			finishTime += increaseTime;
 			ctx.finishTimeSeconds += increaseTime / 6;
 		}
 		double maxTravelTimeCmpToWalk = totalDistance / ctx.cfg.walkSpeed - ctx.cfg.changeTime / 2;
 		List<TransportRouteSegment> results = new ArrayList<TransportRouteSegment>();
 		initProgressBar(ctx, start, end);
+		TLongObjectHashMap<TransportRouteSegment> finalizedSegments = new TLongObjectHashMap<>();
 		while (!queue.isEmpty()) {
 			long beginMs = MEASURE_TIME ? System.currentTimeMillis() : 0;
 			if (ctx.calculationProgress != null && ctx.calculationProgress.isCancelled) {
 				return null;
 			}
 			TransportRouteSegment segment = queue.poll();
 			TransportRouteSegment ex = ctx.visitedSegments.get(segment.getId());
+			TransportRouteSegment finish = null;
 			if(ex != null) {
+				TransportRouteSegment finalized = finalizedSegments.get(segment.getId());
```

Comment:
Technically it's not correct cause segment could be finalized later and this event is independent from that current point is visited